src/app/shared/auth/auth.service.ts

```diff
@@ -29,182 +21,103 @@ export enum AuthServiceState { Idle, Fetching, Refreshing, Error }
-
-    this.subscription = this.accessToken.pipe(pairwise()).subscribe(tokens => this.tokenChanged(tokens));
-
-    // First, check if a code/state is given in URL (i.e., we are back from a oauth login redirect) and try to get a token from it.
+    // (1) check if a code/state is given in URL (i.e., we are back from a oauth login redirect) and try to get a token from it.
     oauthService.tryCompletingCodeFlowLogin().pipe(
-      catchError(_e => of<AccessToken|null>(null)),
-      switchMap(newToken => {
-        const fromStorage = AccessToken.fromStorage();
-
-        // If can get a token from these code/state, use this one in priority
-        if (newToken !== null) {
-          // if there was a stored token, revoke the token
-          if (fromStorage === null) return of(newToken);
-          else return merge(
-            of(newToken),
-            this.authHttp.revokeToken(fromStorage.accessToken).pipe(
-              switchMap(() => of<AccessToken>()) // just completing without emitting anything new
-            )
-          );
+      catchError(_e => {
+        // (2) use the ongoing authentication if any
+        if (appConfig.useTokens) {
+          const token = tokenAuthFromStorage();
+          return token ? of(token) : throwError(new Error('no token stored for token auth'));
+        } else {
+          return this.authHttp.refreshCookie(); // will fail if the browser has no cookie
         }
-
-        // if there is a valid token in storage, use it
-        if (fromStorage !== null) return of(fromStorage);
-
-        // otherwise, create a temp session
-        return this.tempAuth.login().pipe(retry(2));
       }),
+      catchError(_e =>
+        // (3) otherwise, create a temp session
+        this.authHttp.createTempUser().pipe(retry(2))
+      ),
     ).subscribe({
-      next: (token: AccessToken) => {
-        this.accessToken.next(token);
+      next: (auth: AuthResult) => {
+        this.status$.next(auth);
       },
       error: _e => {
         // if temp user creation fails, there is not much we can do
-        this.state$.next(AuthServiceState.Error);
-      },
-      complete: () => {
-        // in any of the success case
-        this.state$.next(AuthServiceState.Idle);
+        this.failure$.next();
       }
     });
-  }
 
-  ngOnDestroy(): void {
-    this.subscription.unsubscribe();
-    this.accessToken.complete();
+    // setup auto-refresh after delay
+    this.status$.pipe(
+      switchMap(auth => {
+        const maxDelay = 2147483647;
+        if (!auth.authenticated) return EMPTY;
+        // Refresh if the token is valid < `minTokenLifetime` or when it will have reached 50% of its lifetime. Retry every minute.
+        const refreshIn = auth.expiration.getTime() - Date.now() <= minTokenLifetime ? 0 :
+          Math.max((auth.expiration.getTime() + auth.creation.getTime())/2 - Date.now(), 0);
```

Comment:
this calculus is not trivial to understand, maybe declaring intermediate variables would increase the readability ?

Reply:
That's not really new to this PR either.
I remember when writing this expression. Actually that's the right-side of ternary operator which is not trivial, but I don't want to create var before L62 for only if we are in the right-side case (this var would not make sense in the other case). So I would need to do it with if/else, but TS does not provide a clean (from my pt of view) way (except ternary operation of separated function) to define an variable conditionally, so at the end alternatives were not more readable... so I ended up with this.